contracts/DivviRegistry.sol

```diff
@@ -0,0 +1,189 @@
+  event RequiresApprovalForRewardsAgreements(
+    address indexed entity,
+    bool requiresApproval
+  );
+  event RewardsAgreementRegistered(
+    address indexed rewardsProvider,
+    address indexed rewardsConsumer
+  );
+
+  // Errors
+  error InvalidEntityAddress(address entity);
+  error EntityAlreadyExists(address entity);
+  error EntityDoesNotExist(address entity);
+  error AgreementAlreadyExists(address provider, address consumer);
+  error ProviderRequiresApproval(address provider);
+
+  /// @custom:oz-upgrades-unsafe-allow constructor
+  constructor() {
+    _disableInitializers();
+  }
+
+  /**
+   * @notice Initialize the contract with an owner and transfer delay
+   * @param owner The address that will have the DEFAULT_ADMIN_ROLE
+   * @param transferDelay The delay in seconds before admin role can be transferred
+   */
+  function initialize(address owner, uint48 transferDelay) public initializer {
+    __AccessControlDefaultAdminRules_init(transferDelay, owner);
+    __UUPSUpgradeable_init();
+  }
+
+  /**
+   * @notice Authorize contract upgrades
+   */
+  function _authorizeUpgrade(
+    address
+  ) internal override onlyRole(DEFAULT_ADMIN_ROLE) {} // solhint-disable-line no-empty-blocks
+
+  /**
+   * @notice Modifier to ensure an entity exists
+   * @param entity The entity address to check
+   */
+  modifier entityExists(address entity) {
+    if (!_entities[entity]) {
+      revert EntityDoesNotExist(entity);
+    }
+    _;
+  }
+
+  /**
+   * @notice Register a new rewards entity
+   * @param entity An address that the entity owns, which will be used to identify the entity and call privileged functions on this contract
+   * @param requiresApproval Whether the entity requires approval for agreements
+   */
+  function registerRewardsEntity(
+    address entity,
+    bool requiresApproval
+  ) external {
+    if (entity == address(0)) {
+      revert InvalidEntityAddress(entity);
+    }
+
+    if (_entities[entity]) {
+      revert EntityAlreadyExists(entity);
+    }
+
+    _entities[entity] = true;
+    _requiresApproval[entity] = requiresApproval;
+    emit RewardsEntityRegistered(entity, requiresApproval);
+  }
+
+  /**
+   * @notice Set whether a Rewards Entity requires approval for agreements
+   * @param requiresApproval Whether the entity requires approval
+   */
+  function setRequiresApprovalForRewardsAgreements(
+    bool requiresApproval
+  ) external entityExists(msg.sender) {
+    _requiresApproval[msg.sender] = requiresApproval;
+    emit RequiresApprovalForRewardsAgreements(msg.sender, requiresApproval);
+  }
+
+  /**
+   * @notice Register a Rewards Consumer - Rewards Provider relationship
+   * @dev Should be called by the Rewards Consumer
+   * @param rewardsProvider The provider entity address
+   */
+  function registerAgreementAsConsumer(
+    address rewardsProvider
+  ) external entityExists(rewardsProvider) entityExists(msg.sender) {
```

Comment:
Quick thought on agreement registration:

The current design doesn't prevent registering an agreement between, say, two entities that might conceptually only be "Consumers".

Is the intention that any registered entity can act as either a Provider or Consumer in an agreement, offering maximum flexibility? Just confirming this flexibility is desired over stricter role enforcement during agreement creation.

Reply:
yes it was documented [here](https://docs.google.com/document/d/12HZQ-e5mWM_oAhKosKg86TdQp0uUPlCZbwI-F-MVoz4/edit?tab=t.0), Joe gave the example where beefy might be a provider and a consumer, where aave might be it's provider. are there downsides to this flexibility? i haven't thought about this a ton but it seems nice that an entity can register once and be both (otherwise they would need to maintain two entities to be both?)

Reply:
Thanks yes makes sense, I like the flexibility.